packages/gatsby-transformer-csv/src/gatsby-node.js

```diff
@@ -25,30 +26,67 @@ async function onCreateNode(
   }
   // Load CSV contents
   const content = await loadNodeContent(node)
+
+  // Destructure out our custom options
+  const { typeName, nodePerFile, ...options } = pluginOptions || {}
+
   // Parse
   let parsedContent = await convertToJson(content, options)
 
+  // Generate the type
+  function getType({ node, object }) {
+    if (pluginOptions && _.isFunction(typeName)) {
+      return pluginOptions.typeName({ node, object })
+    } else if (
+      pluginOptions &&
+      _.isString(typeName) &&
+      typeName === `from-dir`
+    ) {
+      return _.upperFirst(_.camelCase(`${path.basename(node.dir)} Csv`))
+    } else if (
+      pluginOptions &&
+      _.isString(typeName) &&
+      typeName === `from-filename`
+    ) {
+      // be explicit
+      return _.upperFirst(_.camelCase(`${node.name} Csv`))
+    } else {
+      return _.upperFirst(_.camelCase(`${node.name} Csv`))
```

Comment:
Can we do some plugin option validation and applying defaults ahead of doing any processing so we don't have extra code branches here that do similar thing?

For example on line 31 we could have:
```js
let { typeName, nodePerFile, ...options } = pluginOptions || {}
if (nodePerFile && !_.isString(nodePerFile)) {
  // if nodePerFile is truthy, but not a string - let's set default "items"
  nodePerFile = 'items' 
}

if (!typeName) {
  // if typeName is not set, use default:
  typeName = `from-filename`
}
```